src/main/psxsdk/libspu/s_m_f.c

```diff
@@ -1,3 +1,38 @@
 #include "common.h"
+#include "libspu_internal.h"
 
-INCLUDE_ASM("main/nonmatchings/psxsdk/libspu/s_m_f", SpuFree);
+extern s32 D_8003355C;
+
+void SpuFree(unsigned long arg0) {
+    s32 temp_v1;
+    s32 cur_block_num;
+    SPU_MALLOC* cur_mem;
+    s32 temp;
+    s32 mask4;
+    s32 block_num;
+    u8 pad[2];
+    cur_block_num = 0;
+    if (temp_v1 = D_8003355C > 0) {
+        mask4 = 0x40000000;
+        temp = arg0 | 0x80000000;
+        block_num = D_8003355C;
+        cur_mem = _spu_memList;
+        while (true) {
+            temp_v1 = cur_mem->addr;
+
+            if (!(temp_v1 & mask4)) {
+                cur_block_num += 1;
+                if (temp_v1 == arg0) {
+                    cur_mem->addr = temp;
+                } else {
+                    cur_mem++;
+                    if (cur_block_num < block_num) {
+                        continue;
```

Comment:
If you want to remove that `continue` and avoid some nesting, you can rewrite the `while` loop like here: https://decomp.me/scratch/XwfQs

The last `break;` is a bit weird. I tried to convert it into a `for` or a `do{} while` but unsuccessfully.